prune/prune.go

```diff
@@ -16,184 +16,204 @@ package prune
-	// JobKind equates to a Kube Job resource kind
-	JobKind string = "Job"
-	// PodKind equates to a Kube Pod resource kind
-	PodKind string = "Pod"
-)
+/*
+	-------------------------------------------
+	New Auto Pruning API Implementation
+	-------------------------------------------
+*/
+
+// Pruner is an object that runs a prune job.
+type Pruner struct {
+	// Registry is the IsPrunableFunc registry
+	Registry *Registry
+
+	// Client is the k8s client that will be used
+	Client client.Client
+
+	// DryRun indicates whether or not we should actually perform pruning or just return the list of pruneable objects
+	// true = Just check, don't prune
+	// false (default) = Prune
+	DryRun bool
+
+	// Strategy is the function used to determine a list of resources that are pruneable
+	Strategy StrategyFunc
 
-// StrategyConfig holds settings unique to each pruning mode
-type StrategyConfig struct {
-	Mode            Strategy
-	MaxAgeSetting   string
-	MaxCountSetting int
-	CustomSettings  map[string]interface{}
+	// Labels is a map of the labels to use for label matching when looking for resources
+	Labels map[string]string
+
+	// Namespace is the namespace to use when looking for resources
+	Namespace string
+
+	// Logger is the logger to use when running pruning functionality
+	Logger logr.Logger
 }
 
-// StrategyFunc function allows a means to specify
-// custom prune strategies
-type StrategyFunc func(ctx context.Context, cfg Config, resources []ResourceInfo) ([]ResourceInfo, error)
-
-// PreDelete function is called before a resource is pruned
-type PreDelete func(ctx context.Context, cfg Config, something ResourceInfo) error
-
-// Config defines a pruning configuration and ultimately
-// determines what will get pruned
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// ErrUnprunable indicates that it is not allowed to prune a specific object.
+type ErrUnprunable struct {
+	Obj    *client.Object
+	Reason string
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// Error returns a string representation of an `ErrUnprunable` error.
+func (e *ErrUnprunable) Error() string {
+	return fmt.Sprintf("%s -- %+v", e.Reason, e.Obj)
```

Comment:
In general, I like to keep `Error` returns to one line since they will likely be logged, and multiline things make log messages hard to read. This is specifically in regards to the "%+v" format directive for the object, which will dump the whole struct in a way that is pretty hard to use since go's default formatting for complex types is not easy on the eyes.

What I would like is the method here to return a human-readable one-liner:

```go
    return fmt.Sprintf("unable to prune %s: %s", client.ObjectKeyFromObject(e.Obj), e.Reason)
```

`client.ObjectKey` has a `String() string` method  that returns `namespace/name`. The `%s` format directive will automatically call it if present.

Then, where the error actually occurs, you can add a debug log call that dumps the entire object. That way by logs will only have the human-readable description and users can turn on debug logging if they want look at the values of the object. This is a common pattern:

```go
if err != nil {
    log.Debug("a dump of a large struct")
    return err
}
```

Reply:
I agree with the human-readable one-liner idea, so I will work on implementing this.